File: app/lib/core/favorites/favorites_cache.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

/// Lokale Persistenz für Favoriten (Nutzeranfrage Punkt 12, "Offline-
/// Favoriten"/"Konfliktbehandlung"): der zuletzt vom Server bestätigte Stand
/// plus ausstehende, noch nicht synchronisierte Änderungen. Ein Toggle
/// offline schreibt sofort in [pendingChanges] und wirkt lokal sofort —
/// [FavoritesService.toggle] versucht danach den Server-Sync und räumt bei
/// Erfolg aus [pendingChanges] auf.
///
/// Konfliktstrategie: pro Event nur der jeweils LETZTE gewünschte Zustand
/// (`Map<eventId, bool>` statt einer Op-Liste) — mehrfaches Umschalten
/// desselben Events offline kollabiert auf den finalen Wunschzustand statt
/// widersprüchliche Operationen anzuhäufen ("letzter Schreibzugriff
/// gewinnt", die einzig sinnvolle Regel für ein reines Boolean-Favorit ohne
/// serverseitige updated_at-Spalte).
class FavoritesCache {
  const FavoritesCache._();

  static const _knownKey = 'favorites_cache.known_ids';
  static const _pendingKey = 'favorites_cache.pending';

  static Future<Set<String>> knownIds() async {
    final prefs = await SharedPreferences.getInstance();
    return (prefs.getStringList(_knownKey) ?? const []).toSet();
  }

  static Future<void> saveKnownIds(Set<String> ids) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList(_knownKey, ids.toList());
  }

  static Future<Map<String, bool>> pendingChanges() async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_pendingKey);
    if (raw == null) return {};
    try {
      return (jsonDecode(raw) as Map<String, dynamic>).map(
        (k, v) => MapEntry(k, v as bool),
      );
    } catch (_) {
      return {};
    }
  }

  static Future<void> savePendingChanges(Map<String, bool> pending) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_pendingKey, jsonEncode(pending));
  }

  /// Bekannter Serverstand plus ausstehende lokale Änderungen übereinander —
  /// das, was tatsächlich angezeigt werden soll.
  static Future<Set<String>> effectiveIds() async {
    final known = await knownIds();
    final pending = await pendingChanges();
    final effective = {...known};
    for (final entry in pending.entries) {
      if (entry.value) {
        effective.add(entry.key);
      } else {
        effective.remove(entry.key);
      }
    }
    return effective;
  }
}

File: ios-native/KlangradarNative/Core/Repositories/FavoritesCache.swift
import Foundation

/// Lokale Persistenz für Favoriten (Nutzeranfrage Punkt 12, "Offline-
/// Favoriten"/"Konfliktbehandlung") — Pendant zu Flutters FavoritesCache
/// (app/lib/core/favorites/favorites_cache.dart). Der zuletzt vom Server
/// bestätigte Stand plus ausstehende, noch nicht synchronisierte Toggles.
///
/// Konfliktstrategie: pro Event nur der jeweils letzte gewünschte Zustand
/// (`[UUID: Bool]` statt einer Op-Liste) — mehrfaches Umschalten desselben
/// Events offline kollabiert auf den finalen Wunschzustand ("letzter
/// Schreibzugriff gewinnt", die einzig sinnvolle Regel ohne serverseitige
/// updated_at-Spalte auf `favorites`).
enum FavoritesCache {
    private static let knownKey = "favoritesCache.knownIDs"
    private static let pendingKey = "favoritesCache.pending"

    static func knownIDs() -> Set<UUID> {
        let raw = UserDefaults.standard.stringArray(forKey: knownKey) ?? []
        return Set(raw.compactMap(UUID.init(uuidString:)))
    }

    static func saveKnownIDs(_ ids: Set<UUID>) {
        UserDefaults.standard.set(ids.map(\.uuidString), forKey: knownKey)
    }

    static func pendingChanges() -> [UUID: Bool] {
        guard let data = UserDefaults.standard.data(forKey: pendingKey),
              let raw = try? JSONDecoder().decode([String: Bool].self, from: data) else { return [:] }
        return Dictionary(uniqueKeysWithValues: raw.compactMap { key, value in
            UUID(uuidString: key).map { ($0, value) }
        })
    }

    static func savePendingChanges(_ pending: [UUID: Bool]) {
        let raw = Dictionary(uniqueKeysWithValues: pending.map { ($0.key.uuidString, $0.value) })
        guard let data = try? JSONEncoder().encode(raw) else { return }
        UserDefaults.standard.set(data, forKey: pendingKey)
    }

    /// Bekannter Serverstand plus ausstehende lokale Änderungen übereinander.
    static func effectiveIDs() -> Set<UUID> {
        var effective = knownIDs()
        for (id, wantsFavorited) in pendingChanges() {
            if wantsFavorited { effective.insert(id) } else { effective.remove(id) }
        }
        return effective
    }
}
